File: backend/app/schemas/paper_trading.py
"""DTOs for the Paper Trading Command Center.

Everything here is read-only. The API exposes no write endpoint for the
paper engine: only the agent process (paper_trading/) creates decisions
and positions, so no frontend request can ever cause a trade.
"""

from datetime import date, datetime

from pydantic import BaseModel


class PaperDecisionDTO(BaseModel):
    symbol: str
    session_date: date
    decision_timestamp: datetime
    decision: str
    decision_reason: str
    no_trade_reason: str | None = None
    trend_assessment: str
    confidence: int

    strategy_version: str
    configuration_hash: str

    spot: float | None = None
    trend_label: str | None = None
    institutional_bias_label: str | None = None
    vwap: float | None = None
    poc: float | None = None
    support_low: float | None = None
    support_high: float | None = None
    resistance_low: float | None = None
    resistance_high: float | None = None
    atr_14: float | None = None
    rvol_pct: float | None = None
    dominant_side: str | None = None
    pcr: float | None = None
    atm_iv_call: float | None = None
    atm_iv_put: float | None = None

    transition_verdict: str | None = None
    probability_up: float | None = None
    probability_down: float | None = None
    probability_no_move: float | None = None
    expected_move_low: float | None = None
    expected_move_high: float | None = None
    transition_risk_tier: str | None = None
    n_analogs: int | None = None

    option_type: str | None = None
    strike: float | None = None
    expiry: date | None = None
    entry_bid: float | None = None
    entry_ask: float | None = None
    entry_spread_pct: float | None = None
    option_delta: float | None = None
    dynamic_stop: float | None = None
    dynamic_target: float | None = None
    underlying_invalidation: float | None = None
    underlying_target: float | None = None
    reward_risk: float | None = None
    quantity: int | None = None
    capital_allocated: float | None = None

    supporting_factors: list[str] = []
    conflicting_factors: list[str] = []
    risk_factors: list[str] = []
    explanation: list[str] = []
    data_quality: str
    missing_fields: list[str] = []

    model_config = {"from_attributes": True}


class PaperPositionEventDTO(BaseModel):
    event_timestamp: datetime
    minutes_in_trade: int
    event_type: str
    underlying_price: float | None = None
    option_bid: float | None = None
    option_ask: float | None = None
    unrealized_pnl: float | None = None
    momentum: str | None = None
    current_stop: float | None = None
    current_target: float | None = None
    note: str | None = None

    model_config = {"from_attributes": True}


class PaperPositionDTO(BaseModel):
    id: int
    symbol: str
    session_date: date
    option_type: str
    strike: float
    expiry: date | None = None
    quantity: int
    entry_timestamp: datetime
    entry_bid: float | None = None
    entry_ask: float | None = None
    entry_spread_pct: float | None = None
    entry_price: float
    capital_allocated: float
    spot_at_entry: float | None = None
    initial_stop: float
    initial_target: float
    current_stop: float
    current_target: float
    is_open: bool
    exit_timestamp: datetime | None = None
    exit_bid: float | None = None
    exit_price: float | None = None
    exit_reason: str | None = None
    gross_pnl: float | None = None
    net_pnl: float | None = None
    pnl_basis: str
    return_pct: float | None = None
    mfe_pct: float | None = None
    mae_pct: float | None = None
    underlying_move: float | None = None
    direction_correct: bool | None = None
    data_quality: str
    events: list[PaperPositionEventDTO] = []

    model_config = {"from_attributes": True}


class PaperAccountDTO(BaseModel):
    session_date: date | None = None
    starting_capital: float
    current_capital: float
    available_capital: float
    capital_in_trade: float
    realized_pnl: float
    unrealized_pnl: float
    daily_pnl: float
    total_pnl: float
    daily_drawdown: float
    max_drawdown: float
    max_drawdown_pct: float
    trade_count: int
    win_count: int
    loss_count: int
    win_rate: float | None = None
    consecutive_wins: int
    consecutive_losses: int
    average_win: float | None = None
    average_loss: float | None = None
    largest_win: float | None = None
    largest_loss: float | None = None
    profit_factor: float | None = None
    pnl_basis: str = "PRE_COST"


class PaperStatusDTO(BaseModel):
    paper_mode: bool
    agent_state: str
    kill_switch_active: bool
    session_date: date | None = None
    last_heartbeat: datetime | None = None
    strategy_version: str
    configuration_version: str
    configuration_hash: str
    experiment_day: int | None = None
    experiment_total_days: int = 5
    underlying_data_age_sec: float | None = None
    option_data_age_sec: float | None = None
    database_ok: bool = True


class DecisionOutcomeDTO(BaseModel):
    """Direction correctness and option profitability are separate
    questions -- a correct call that lost money is not a good trade, and
    a wrong call that made money is not evidence of skill."""
    symbol: str
    session_date: date
    decision: str
    forecast_direction: str | None = None
    actual_direction: str | None = None
    direction_correct: bool | None = None
    option_profitable: bool | None = None
    net_pnl: float | None = None
    decision_quality: str


class PaperTodayDTO(BaseModel):
    session_date: date
    status: PaperStatusDTO
    account: PaperAccountDTO
    decisions: list[PaperDecisionDTO]
    open_position: PaperPositionDTO | None = None
    closed_today: list[PaperPositionDTO] = []
    outcomes: list[DecisionOutcomeDTO] = []

File: frontend/src/types/paperTrading.ts
// Mirrors backend/app/schemas/paper_trading.py (hand-synced, matching this
// project's existing convention -- there is no shared codegen).
// PAPER MODE: every value here describes a simulated experiment. No real
// order is ever placed and all P&L is PRE-COST.

export interface PaperStatusDTO {
  paper_mode: boolean;
  agent_state: string;
  kill_switch_active: boolean;
  session_date: string | null;
  last_heartbeat: string | null;
  strategy_version: string;
  configuration_version: string;
  configuration_hash: string;
  experiment_day: number | null;
  experiment_total_days: number;
  underlying_data_age_sec: number | null;
  option_data_age_sec: number | null;
  database_ok: boolean;
}

export interface PaperAccountDTO {
  session_date: string | null;
  starting_capital: number;
  current_capital: number;
  available_capital: number;
  capital_in_trade: number;
  realized_pnl: number;
  unrealized_pnl: number;
  daily_pnl: number;
  total_pnl: number;
  daily_drawdown: number;
  max_drawdown: number;
  max_drawdown_pct: number;
  trade_count: number;
  win_count: number;
  loss_count: number;
  win_rate: number | null;
  consecutive_wins: number;
  consecutive_losses: number;
  average_win: number | null;
  average_loss: number | null;
  largest_win: number | null;
  largest_loss: number | null;
  profit_factor: number | null;
  pnl_basis: string;
}

export interface PaperDecisionDTO {
  symbol: string;
  session_date: string;
  decision_timestamp: string;
  decision: "TRADE_CALL" | "TRADE_PUT" | "NO_TRADE";
  decision_reason: string;
  no_trade_reason: string | null;
  trend_assessment: string;
  confidence: number;
  strategy_version: string;
  configuration_hash: string;
  spot: number | null;
  trend_label: string | null;
  institutional_bias_label: string | null;
  vwap: number | null;
  poc: number | null;
  support_low: number | null;
  support_high: number | null;
  resistance_low: number | null;
  resistance_high: number | null;
  atr_14: number | null;
  rvol_pct: number | null;
  dominant_side: string | null;
  pcr: number | null;
  atm_iv_call: number | null;
  atm_iv_put: number | null;
  transition_verdict: string | null;
  probability_up: number | null;
  probability_down: number | null;
  probability_no_move: number | null;
  expected_move_low: number | null;
  expected_move_high: number | null;
  transition_risk_tier: string | null;
  n_analogs: number | null;
  option_type: string | null;
  strike: number | null;
  expiry: string | null;
  entry_bid: number | null;
  entry_ask: number | null;
  entry_spread_pct: number | null;
  option_delta: number | null;
  dynamic_stop: number | null;
  dynamic_target: number | null;
  underlying_invalidation: number | null;
  underlying_target: number | null;
  reward_risk: number | null;
  quantity: number | null;
  capital_allocated: number | null;
  supporting_factors: string[];
  conflicting_factors: string[];
  risk_factors: string[];
  explanation: string[];
  data_quality: string;
  missing_fields: string[];
}

export interface PaperPositionEventDTO {
  event_timestamp: string;
  minutes_in_trade: number;
  event_type: string;
  underlying_price: number | null;
  option_bid: number | null;
  option_ask: number | null;
  unrealized_pnl: number | null;
  momentum: string | null;
  current_stop: number | null;
  current_target: number | null;
  note: string | null;
}

export interface PaperPositionDTO {
  id: number;
  symbol: string;
  session_date: string;
  option_type: string;
  strike: number;
  expiry: string | null;
  quantity: number;
  entry_timestamp: string;
  entry_bid: number | null;
  entry_ask: number | null;
  entry_spread_pct: number | null;
  entry_price: number;
  capital_allocated: number;
  spot_at_entry: number | null;
  initial_stop: number;
  initial_target: number;
  current_stop: number;
  current_target: number;
  is_open: boolean;
  exit_timestamp: string | null;
  exit_bid: number | null;
  exit_price: number | null;
  exit_reason: string | null;
  gross_pnl: number | null;
  net_pnl: number | null;
  pnl_basis: string;
  return_pct: number | null;
  mfe_pct: number | null;
  mae_pct: number | null;
  underlying_move: number | null;
  direction_correct: boolean | null;
  data_quality: string;
  events: PaperPositionEventDTO[];
}

export interface DecisionOutcomeDTO {
  symbol: string;
  session_date: string;
  decision: string;
  forecast_direction: string | null;
  actual_direction: string | null;
  direction_correct: boolean | null;
  option_profitable: boolean | null;
  net_pnl: number | null;
  decision_quality: string;
}

export interface PaperTodayDTO {
  session_date: string;
  status: PaperStatusDTO;
  account: PaperAccountDTO;
  decisions: PaperDecisionDTO[];
  open_position: PaperPositionDTO | null;
  closed_today: PaperPositionDTO[];
  outcomes: DecisionOutcomeDTO[];
}
